### PR TITLE
LossCompute rework

### PR DESCRIPTION
**Pull request by JianyuZhan**
This PR reworks the `*LossCompute`  interfaces,  mainly addressing:

- removed the `forward()` interface from the base class.
- removed `copy_attn` knowledge from the base class. This moves the `make_shard_state()` interface to each subclass.  The subclass implements its own `compute_loss()`, whose interface corresponds to the return values of `make_shard_state()`, so it makes sense to implement `make_shard_state()` for each subclass.
- added a `monolithic_compute_loss()` for computing the loss monolithically, compared to `sharded_compute_loss()`, which computes the loss in shards.
- removed the `**kwargs` arg in the subclass's  `compute_loss()` interface. This make the subclass's interface concrete and explicit for what they want.

Now the new interface in `Loss.py` is totally agnostic to `copy_attn`,  and all `copy_attn` knowledge are hided in `onmt/module/CopyGenerater.py`. And the `make_shard_state()` is no more cluttered with `if...else...`.

Detailed documentations are added for each interface in the base class,  which would be clear for understanding and helps future work.

New `LossComputeBase` interface:
https://github.com/JianyuZhan/OpenNMT-py/blob/loss_compute_rework/onmt/Loss.py#L15

New `NMTLossCompute` interface:
https://github.com/JianyuZhan/OpenNMT-py/blob/loss_compute_rework/onmt/Loss.py#L101

New `CopyGeneratorCompute` interface:
https://github.com/JianyuZhan/OpenNMT-py/blob/loss_compute_rework/onmt/modules/CopyGenerator.py#L79

**Test done**:
tests w/  and w/o  `copy_attn` have been run.

### CODE REVIEW

**Comment by bpopeters** [comment]
I'm not sure that these `LossCompute` classes need to be submodules of `nn.Module`. The point of an `nn.Module` is that you can define various parameters in `__init__`, the behavior of the forward pass in `forward`, and the superclass will take care of the rest without trying to worry about it too much. The provided justification is to use cuda, but isn't that unnecessary because we'll have already done the cuda stuff elsewhere, with the objects that the `LossCompute` objects wrap around.

In fact, I'm not sure that computing the loss requires a class at all. There is no interesting state here, just a way to bundle together the loss function, the sharding behavior, the back-propagation of the loss, and statistics-gathering into one class. It might be that packaging this stuff in a class is the best way to make it easy to extend if someone wanted to use a different loss function or something, but I think alternatives deserve exploration. 

Also, is it necessary and desirable to log stats or execute the backward pass inside a function and class whose stated purpose is to compute the loss? It seems like a good idea to me to make these tasks as modular as possible.

**Comment by JianyuZhan** [comment]
Hi, @bpopeters,

1.  True that the only justification for using `nn.Module` is for cuda behavior.  The loss compute duty is independent of the model itself. We indeed make the whole model  handle the cuda stuff, but that model hierarchy(by subclassing `nn.Module` too) is as far beyond as this gpu behavior could propagate.  Because the `Trainer` object composes of the model and the loss, among the other things(we try to make trainer modular),  so we need explicitly handle cuda for loss as well(previously, it is an opt passed into the `NMTCriterion`).  And you are right using `nn.Module` is a bit abuse,  but since we don't want to stuff the model with too many things, this is a compromise.

2.  To make the LossCompute as a hierarchical class is not because the state en-capsule justification, it is just because we want to tighten up the trainer interface and also make it flexible to accept a **callable** object for computing loss(what to do), but agnostic to how it is computed(how to do it),  be it a class, a function, or a lambda.  And I found out that bundle these seemly irrelevant things into a LossCompute class does help clarify its interface. For example, the pending `Extended Loss`  PR could benefit from this hierarchy too and does little things itself while obeying with the interface protocol. I deem this as the power and simplicity of dividing mechanism from strategy -  clean interface and hide irrelevant details.  We could split these things- the loss computing function, the shard behavior, the stats, but eventually they still need to be passed to the trainer somehow, which would make the trainer interface fragile.

**Comment by srush** [comment]
This PR is an improvement, so I will merge it. 

However, I partially agree with @bpopeters somewhat.

1) I am fine with it being classes, as there may be future losses that require parameter state. I think @bpopeters is too anti-class. There are plenty of nn modules that have no parameters. 

2) I don't like that we compute the loss and the statistics together in forward. That may have been my fault originally. Forward has a very precise interface. It should take in variables and return variables. Statistics should be computed outside somehow. That is independent of these changes.

**Comment by bpopeters** [comment]
I can understand why one would think I am anti-class, but I don't think it's warranted. Python is a multi-paradigm language that doesn't force us into one way of thinking, and that's a good thing. Classes are often the right solution, but that doesn't mean they always are. I think there are two distinct issues here related to classes.

1.  It might be true that a class is the right way to compute loss, but I am not completely convinced by the arguments I've seen so far. @JianyuZhan said that the goal is for the trainer to be agnostic about the callable object for computing loss, but if the trainer is agnostic on this point, there's no reason for us to favor a class here unless we have a good reason. @srush's note that other ways to compute loss may require parameter state seems like a good reason. The attributes currently defined here, on the other hand, seem likely to cause more harm than good. Why are `self.tgt_vocab` and `self.padding_idx` defined? The only class-internal use of `self.padding_idx` is inside the `stats` method, which for already stated reasons probably belongs somewhere else. `self.tgt_vocab` is never used class-internally. It doesn't seem plausible to me that anything else in the code would ever need to access or mutate those attributes. There is no documentation to clear these points up. A couple of the attributes are beneficial, of course. The `self.criterion` obviously does the loss computing that this class wraps around, and the `self.generator` is necessary until we come up with a different way to efficiently run what is, in a conceptual sense, the output layer of the decoder (see #314; sounds like it could take a while to do this).

2. The second point concerns the use of `nn.Module`. I agree that not all `nn.Module` subclasses define parameter state. This is the case for modules for non-linear activation functions, for example. But those modules are just wrappers around functions and are defined so that activation functions can be used inside `nn.Sequential`: it's just a wrapper. That's a pretty good reason to subclass `nn.Module`. But that's not what's going on here. @JianyuZhan's reason was to make use of cuda behavior even though the loss compute objects do not use the `nn.Module` interface and are conceptually different from typical modules. The gpu optimizations can be done explicitly instead of implicitly, and explicit is better than implicit.
